### DISCUSSION TITLE
Solved: Debugging into hardware implementations (e.g. ScanningProbeInterface)

### DISCUSSION THREAD

**Discussion by gemmell** [Ulm-IQO/qudi-iqo-modules#86] [Q&A] [answered]
Hi,

I haven't quite got my head around what's going on, so I'm hoping someone here has already experienced this. I am putting debug breakpoints in `scanning_probe_dummy.py` and the breakpoints are hit in `__init__` and `get_constraints`, but functions like `configure_scan` and `start_scan` don't break on the breakpoints, even though I know it's getting there (I can see log messages I put in there).

It seems like it breaks if it's called from the main thread, but doesn't break if it's called from the UI thread? Has anyone come across this and more to the point: does anyone know the trick to getting it to hit those breakpoints? It's not being called from a separate `process` is it? 

I'm using visual studio code rather pycharm, but that shouldn't make any difference - it's obviously able to get to the symbols & debug through the file, it's just that the mechanism by which the UI calls through to those functions somehow bypasses the debugger...

Any ideas?

**Comment by gemmell** [comment]
It's a problem with vscode! More info here:
https://stackoverflow.com/a/72471422

And copying from the Microsoft site linked there:
> If you're working with a multi-threaded app that uses native thread APIs (such as the Win32 CreateThread function rather than the Python threading APIs), it's presently necessary to include the following source code at the top of whichever file you want to debug:
```python
import debugpy
debugpy.debug_this_thread()
```

**Comment by timoML** [comment]
Does this allow you to debug from GUI with Vscode?
My experience with Pycharm is:
- From GUI, it's pretty random whether a breakpoint is hit. Most of the times, it won't though.
- My workaround so far was to call the function that should be debugged from the qudi manager. Then, the breakpoint would be hit relatively reliably.

**Reply by gemmell** [reply, marked as answer]
Yeah, no problems once you do this:
`import debugpy`
add `debugpy.debug_this_thread()` to the top of all the functions in your device. e.g. for the scanner_probe I'm implementing it's:
```python
 def configure_scan(self, scan_settings):
        debugpy.debug_this_thread()

        with self._thread_lock:
           ...

def start_scan(self):
        debugpy.debug_this_thread()

        with self._thread_lock:
           ...
 ```
 
 It is 100% reliable hitting breakpoints after this.
